timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosQuorumCheckingCoalescingFunction.java

```diff
@@ -96,27 +108,28 @@ public PaxosQuorumCheckingCoalescingFunction(
     public static <REQ, RESP extends PaxosResponse, SERVICE, F extends CoalescingRequestFunction<REQ, RESP>>
     PaxosQuorumCheckingCoalescingFunction<REQ, RESP, F> wrapWithRemotes(
             List<SERVICE> services,
-            ExecutorService executor,
+            Map<SERVICE, ExecutorService> executors,
             int quorumSize,
             Function<SERVICE, F> functionFactory) {
-        return services.stream()
-                .map(functionFactory)
-                .collect(collectingAndThen(
-                        toList(), functions -> new PaxosQuorumCheckingCoalescingFunction<>(
-                                functions,
-                                Maps.toMap(functions, $ -> executor),
-                                quorumSize)));
+        // Not 100% sure if ordering is important, but assuming so here
+        List<FunctionAndExecutor<F>> functionsAndExecutors = KeyedStream.of(services)
+                .map(executors::get)
+                .mapKeys(functionFactory)
+                .entries()
+                .<FunctionAndExecutor<F>>map(entry -> ImmutableFunctionAndExecutor.of(entry.getKey(), entry.getValue()))
+                .collect(Collectors.toList());
+        return new PaxosQuorumCheckingCoalescingFunction<>(functionsAndExecutors, quorumSize);
```

Comment:
I think we went a bit overboard with fanciness to keep it fluent as opposed to
```suggestion
        List<F> functions = new ArrayList<>(services.size());
        Map<F, ExecutorService> executorMap = new HashMap<>(services.size());
        for (SERVICE service: services) {
            F function = functionFactory.apply(service);
            functions.add(function);
            executorMap.put(function, executors.get(service));
        }
        return new PaxosQuorumCheckingCoalescingFunction<>(functions, executorMap, quorumSize);```

Reply:
Fair enough, when I threaded the needle I (probably wrongly) thought it was clear enough 😅 but agree this version is clearer.